File: 101-109_small_problems/easy_7/multiply_lists.rb
# Multiply Lists
# Write a method that takes two Array arguments in which each Array contains a list of numbers, and returns a new Array that contains the product of each pair of numbers from the arguments that have the same index. You may assume that the arguments contain the same number of elements.

# Input: two array arguments
# Output: New array
# Mental model: Given two arrays, multiply the elements of each array that have the same index together.

# Examples:

# multiply_list([3, 5, 7], [9, 10, 11]) == [27, 50, 77]

# Data structures

# Input is two arrays. Output is one array.

# Algorithm
# Create new array
# Loop through first array
# Loop through second array
#   Multiply elements together
#   Assign product to new array
# Return new array

# Code
def multiply_list(arr1, arr2)
  products = []
  counter = 0

  loop do
    products << (arr1[counter] * arr2[counter])

    counter += 1
    break if counter == arr1.length
  end

  products
end

p multiply_list([3, 5, 7], [9, 10, 11]) == [27, 50, 77]

# Further Exploration
# The Array#zip method can be used to produce an extremely compact solution to this method. Read the documentation for zip, and see if you can come up with a one line solution (not counting the def and end lines).

def multiply_list(arr1, arr2)
  arr1.zip(arr2).map {|a, b| a * b }
end

p multiply_list([3, 5, 7], [9, 10, 11]) == [27, 50, 77]

File: exercises/js210_small_problems/easy_4/multiplyLists.js
// Multiply Lists
// Write a function that takes two array arguments, each containing a list of numbers, and returns a new array that contains the product of each pair of numbers from the arguments that have the same index. You may assume that the arguments contain the same number of elements.

// Input: Two array arguments each containing numbers
// Output: A new array containing the product of each pair

// Rules:
// - each input array has the same number of elements
// - elements that have the same index form a pair
// - each number in pair is multiplied together to form the numbers in the new array
//
// Data Structure
// arrays
//
// Algorithm
// Create new array
// Loop over first array
//  Multiply current element from first array * current element from second array
// Add result of multiplication to new array
// Return new array

function multiplyList(firstArray, secondArray) {
  var multiplied = [];
  var result;
  var i;

  for (i = 0; i < firstArray.length; i += 1) {
    result = firstArray[i] * secondArray[i];
    multiplied.push(result);
  }

  return multiplied;
}

// Example:

console.log(multiplyList([3, 5, 7], [9, 10, 11]));    // [27, 50, 77]
